dataset/bucket.cpp

```diff
@@ -183,6 +183,32 @@ auto concatenate_impl(const VariableConstView &var0,
   return Variable{
       std::make_unique<variable::DataModel<bucket<T>>>(combine<T>(var0, var1))};
 }
+
+template <class T>
+auto concatenate_impl(const VariableConstView &var, const Dim dim) {
+  const auto &[indices, buffer_dim, buffer] = var.constituents<bucket<T>>();
+  const auto [begin, end] = unzip(indices);
+  const auto sizes = end - begin;
+  const auto out_sizes = sum(sizes, dim);
+  const auto [out_begin, size] = sizes_to_begin(out_sizes);
+  const auto out_end = out_begin + out_sizes;
+  auto out_buffer = resize_buffer(buffer, dim, size);
+  const auto nslice = indices.dims()[dim];
+  auto out_current = out_begin;
+  auto out_next = out_current;
+  for (scipp::index i = 0; i < nslice; ++i) {
+    const auto slice_indices = indices.slice({dim, i});
+    const auto [slice_begin, slice_end] = unzip(slice_indices);
+    out_next += slice_end;
+    out_next -= slice_begin;
```

Comment:
Here I don't quite follow exactly what's going on. Can you explain a little? (either in a comment in the code or just as a reply to this comment).

Reply:
Will do. This is a draft anyway at this point, since mask handling is still missing.